server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,539 @@
+/**
+ * Friends list plugin.
+ * Allows for adding and removing friends, as well as seeing their activity.
+ * Written by Mia.
+ * @author mia-pi-git
+ */
+
+import {Utils} from '../../lib/utils';
+import {MAX_REQUESTS, sendPM} from '../friends';
+
+const STATUS_COLORS: {[k: string]: string} = {
+	idle: '#ff7000',
+	online: '#009900',
+	busy: '#cc3838',
+};
+
+const STATUS_TITLES: {[k: string]: string} = {
+	online: 'Online',
+	idle: 'Idle',
+	busy: 'Busy',
+	offline: 'Offline',
+};
+
+export const Friends = new class {
+	async notifyPending(user: User) {
+		if (user.settings.blockFriendRequests) return;
+		const friendRequests = await Chat.Friends.getRequests(user);
+		const pendingCount = friendRequests.received.size;
+		if (pendingCount < 1) return;
+		sendPM(`/nonotify You have ${pendingCount} friend requests pending!`, user.id);
+		sendPM(`/raw <button class="button" name="send" value="/j view-friends-received">View</button></div>`, user.id);
+	}
+	async notifyConnection(user: User) {
+		const connected = await Chat.Friends.getLastLogin(user.id);
+		if (connected && (Date.now() - connected) < 2 * 60 * 1000) {
+			return;
+		}
+		const friends = await Chat.Friends.getFriends(user.id);
+		const message = `/nonotify Your friend ${Utils.escapeHTML(user.name)} has just connected!`;
+		for (const f of friends) {
+			const {user1, user2} = f;
+			const friend = user1 !== user.id ? user1 : user2;
+			const curUser = Users.get(friend as string);
+			if (curUser?.settings.allowFriendNotifications) {
+				curUser.send(`|pm|&|${curUser.getIdentity()}|${message}`);
+			}
+		}
+	}
+	async visualizeList(userid: ID) {
+		const friends = await Chat.Friends.getFriends(userid);
+		const categorized: {[k: string]: string[]} = {
+			online: [],
+			idle: [],
+			busy: [],
+			offline: [],
+		};
+		const loginTimes: {[k: string]: number} = {};
+		for (const {friend: friendID, last_login, allowing_login: hideLogin} of [...friends].sort()) {
+			const friend = Users.get(friendID);
```

Comment:
Won't this show users' alts? I usually use alts to get away from people.

Reply:
No, it won't. In displayFriend it checks getExact, so if they're on an alt, it says they're on an alt (but not the name of said alt). This is the same info you can get with /user.